### PR TITLE
Comply with content types for the accept header

### PR DESCRIPTION
**Pull request by 82marbag**
Comply with content-types as described in the documentation [0]

[0] https://awslabs.github.io/smithy/1.0/spec/aws/aws-restjson1-protocol.html#content-type

## Motivation and Context

Comply with the spec and pass these two tests: RestJsonWithPayloadExpectsModeledAccept and RestJsonWithBodyExpectsApplicationJsonAccept

In the next PR: add a test to fix RestJsonWithPayloadExpectsImpliedAccept which is currently broken in smithy

## Description

The accept header should be
* application/json if there's no `@httpPayload`
* `@mediaType`'s value if it exists
* accept any Blob
* the expected content type for the shapes otherwise

Use `NotAcceptable` (406) for requests that don't comply with this.

## Testing

`./gradlew :codegen-server-test:test && ./gradlew :codegen-test:test`

## Checklist

- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt [hunk 1]
```diff
@@ -159,6 +159,22 @@ private class ServerHttpBoundProtocolTraitImplGenerator(
         val operationName = symbolProvider.toSymbol(operationShape).name
         val inputName = "${operationName}${ServerHttpBoundProtocolGenerator.OPERATION_INPUT_WRAPPER_SUFFIX}"
 
+        val impliedResponseContentType = httpBindingResolver.responseContentType(operationShape)
+        val verifyResponseContentType = if (impliedResponseContentType == null) { "" } else {
+            """
+            if let Some(headers) = req.headers() {
+                if let Some(accept) = headers.get(#{http}::header::ACCEPT) {
+                    if accept != ${impliedResponseContentType.dq()} {
```
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:167]:
```suggestion
                    if accept != "${impliedResponseContentType}" {
```

`.dq` is useful when you're interpolating within single double quotes (`""`).

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/RestJson.kt [hunk 2]
```diff
@@ -56,15 +62,32 @@ class RestJsonFactory : ProtocolGeneratorFactory<HttpBoundProtocolGenerator> {
...
      * overridden by a specific mechanism e.g. an output shape member is targeted with `httpPayload` or `mediaType` traits.
      */
-    override fun responseContentType(operationShape: OperationShape): String =
-        super.responseContentType(operationShape) ?: "application/json"
+    override fun responseContentType(operationShape: OperationShape): String? {
+        val members = operationShape
+            .outputShape(model)
+            .members()
+        // temporary fix for https://github.com/awslabs/smithy/blob/df456a514f72f4e35f0fb07c7e26006ff03b2071/smithy-model/src/main/java/software/amazon/smithy/model/knowledge/HttpBindingIndex.java#L352
+        // https://github.com/awslabs/smithy/issues/1259
```
Comment (david-perez) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/RestJson.kt:77]:
```suggestion
        // Temporary fix for https://github.com/awslabs/smithy/blob/df456a514f72f4e35f0fb07c7e26006ff03b2071/smithy-model/src/main/java/software/amazon/smithy/model/knowledge/HttpBindingIndex.java#L352
        // https://github.com/awslabs/smithy/issues/1259
```

TODOs let us grep through the codebase later.

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/RestJson.kt [hunk 3]
```diff
@@ -56,15 +62,32 @@ class RestJsonFactory : ProtocolGeneratorFactory<HttpBoundProtocolGenerator> {
...
+        for (m in members) {
+            if (m.hasTrait<DocumentationTrait>()) {
+                break
+            }
```
Comment (david-perez) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/RestJson.kt:81]:
```suggestion
```

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/RestJson.kt [hunk 4]
```diff
@@ -56,15 +62,32 @@ class RestJsonFactory : ProtocolGeneratorFactory<HttpBoundProtocolGenerator> {
...
      */
-    override fun responseContentType(operationShape: OperationShape): String =
-        super.responseContentType(operationShape) ?: "application/json"
+    override fun responseContentType(operationShape: OperationShape): String? {
+        val members = operationShape
+            .outputShape(model)
+            .members()
+        // temporary fix for https://github.com/awslabs/smithy/blob/df456a514f72f4e35f0fb07c7e26006ff03b2071/smithy-model/src/main/java/software/amazon/smithy/model/knowledge/HttpBindingIndex.java#L352
+        // https://github.com/awslabs/smithy/issues/1259
+        for (m in members) {
```
Comment (david-perez) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/RestJson.kt:78]:
```suggestion
        for (member in members) {
```

**Review by david-perez — APPROVED** [review]

File: rust-runtime/aws-smithy-http-server/src/runtime_error.rs [hunk 5]
```diff
@@ -48,6 +48,7 @@ impl RuntimeErrorKind {
             RuntimeErrorKind::Serialization(_) => "SerializationException",
             RuntimeErrorKind::InternalFailure(_) => "InternalFailureException",
             RuntimeErrorKind::UnknownOperation => "UnknownOperation",
```
[the remark this answers is not in the document]
Comment (crisidev) [reply rust-runtime/aws-smithy-http-server/src/runtime_error.rs:50]: I think it is worth it to fix it in this PR. Thoughts?

[the remark this answers is not in the document]
Comment (82marbag) [reply rust-runtime/aws-smithy-http-server/src/runtime_error.rs:50]: Yes, changing

**Review by crisidev — APPROVED** [review]

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt [hunk 6]
```diff
@@ -159,6 +159,26 @@ private class ServerHttpBoundProtocolTraitImplGenerator(
         val operationName = symbolProvider.toSymbol(operationShape).name
         val inputName = "${operationName}${ServerHttpBoundProtocolGenerator.OPERATION_INPUT_WRAPPER_SUFFIX}"
 
+        val verifyResponseContentType: Writable = writable {
```
Comment (rcoh) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:162]: nit: you don't need `:Writable` here

**Review by rcoh — APPROVED** [review]

**Pull request merged**